opsdroid/events.py

```diff
@@ -0,0 +1,166 @@
+        """Make opsdroid wait x-seconds before responding.
+
+        Number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('thinking-delay', 0)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(seconds)
+
+    async def _typing_delay(self, text):
+        """Delays reply to simulate typing.
+
+        Seconds to delay equals number of characters in response multiplied by
+        number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('typing-delay', 0)
+        char_count = len(text)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(char_count*seconds)
+
+    async def respond(self, response_event, room=None):
+        """Respond to this message using the connector it was created by.
+
+        Creates copy of this message with updated text as response.
+        Delays message if thinking or typing delay present in config. file.
+        Updates responded_to attribute to True if False.
+        Logs response and response time in OpsDroid object stats.
+        """
+        opsdroid = get_opsdroid()
+        if isinstance(response_event, str):
+            response = copy(self)
+            response.text = response_event
+        else:
+            response = response_event
+
+        if 'thinking-delay' in self.connector.configuration or \
+           'typing-delay' in self.connector.configuration:
+            await self._thinking_delay()
+            await self._typing_delay(response.text)
+
+        await self.connector.respond(response, room)
+        if not self.responded_to:
+            now = datetime.now()
+            opsdroid.stats["total_responses"] = \
+                opsdroid.stats["total_responses"] + 1
+            opsdroid.stats["total_response_time"] = \
+                opsdroid.stats["total_response_time"] + \
+                (now - self.created).total_seconds()
+            self.responded_to = True
+
+    async def react(self, emoji):
+        """React to this message with emoji using the specified connector.
+
+        Delays message if thinking delay present in config. file.
+
+        Args:
+            emoji: Sting name of emoji with which OpsDroid will react.
+
+        Returns:
+            bool: True for message successfully sent. False otherwise.
+
+        """
+        if 'thinking-delay' in self.connector.configuration:
+            await self._thinking_delay()
+        return await self.connector.react(self, emoji)
+
+
+class File(Event):
+    """Event class to represent arbitrary files as bytes."""
+
+    def __init__(self, user, room, connector,
+                 file_bytes, url=None):  # noqa: D107
+        super().__init__(user, room, connector)
+        self.file_bytes = file_bytes
```

Comment:
I would suggest we make `file_bytes` a property which can either return `self._file_bytes` if it is set or call a method to download the bytes from the url.

Reply:
Makes sense. Again, I might leave this for the upgrade rather than including it now, unless you have suggestions for a quick and general way of implementing the download bit.

Reply:
Happy to bump for now.